File: Problem10.java
// Time Complexity : O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Problem10 {
    public int findMin(int[] nums) {
        int min = 5001;
        int left = 0;
        int right = nums.length - 1;
        while(left<=right){
            int mid = left + (right - left)/2;
            //check if min value is greater than mid value
            //if it is greater change the min to mid value
            if(min>nums[mid]){
                min = nums[mid];
            }
            //Then check for sorted side in the array
            //if left side is sorted
            if(nums[left]<=nums[mid]){
                //if min value is greater than left value 
                //assign left value to min
                if(min>nums[left]){
                    min = nums[left];
                }
                //And eliminate left half
                left = mid + 1; // Search in the right half
            }else{
                //means right side is sorted
                //if min value is greater than mid value 
                //assign mid value to min
                if(min>nums[mid]){
                    min = nums[mid];
                }
                //And eliminate right half
                right = mid - 1; // Search in the left half
            }
        }
        return min;
    }
}

File: problem 10.py
# Time Complexity : O(log n)
# Space Complexity :o(1)
# Did this code successfully run on Leetcode : YES
# Any problem you faced while coding this : NO


# Your code here along with comments explaining your approach in three sentences only

# in the rotated sorted array if the low <= high then the min is at low or when the mid is calculated 
# and the mid itself happens to be the min than their neighbors then that is the min
# otherwise the min will be in unsorted side of the mid in the array and we go that way by performing BS.



class Solution:
    def findMin(self, nums: List[int]) -> int:
        low = 0
        high = len(nums) - 1
        while(low <= high):
            if nums[low] <= nums[high]: # this means the array is not rotated or rotated n times 
                return nums[low] # or that part of the array has its min at low so return that
            mid = low + (high - low) // 2
            if nums[mid - 1] > nums[mid] and nums[mid] < nums[mid + 1]: # if mid itself is least than its neighbors
                return nums[mid] # then return mid
            elif nums[low] <= nums[mid]: # if the array is sorted on the left then go to right
                low = mid+1 # as the min will be in the unsorted side
            else:
                high = mid - 1 # otherwise go to left